# IBM Applied Data Science Capstone Course by Coursera
### Week 5 Final Report
**_Opening a New Coffee Shop in Kuala Lumpur, Malaysia_**
- Build a dataframe of neighborhoods in Kuala Lumpur, Malaysia by web scraping the data from Wikipedia page
- Get the geographical coordinates of the neighborhoods
- Obtain the venue data for the neighborhoods from Foursquare API
- Explore and cluster the neighborhoods
- Select the best cluster to open a new shopping mall
***
### 1. Import libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import geocoder # to get coordinates

import requests # library to handle requests
from bs4 import BeautifulSoup # library to parse HTML and XML documents

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print("Libraries imported.")

Libraries imported.


### 2. Scrap data from Wikipedia page into a DataFrame

In [2]:
# send the GET request
data = requests.get("https://en.wikipedia.org/wiki/Category:Suburbs_in_Kuala_Lumpur").text

In [3]:
# parse data from the html into a beautifulsoup object
soup = BeautifulSoup(data, 'html.parser')

In [4]:
# create a list to store neighborhood data
neighborhoodList = []

In [5]:
# append the data into the list
for row in soup.find_all("div", class_="mw-category")[0].findAll("li"):
    neighborhoodList.append(row.text)

In [6]:
# create a new DataFrame from the list
kl_df = pd.DataFrame({"Neighborhood": neighborhoodList})

kl_df.head()

Neighborhood
0             Alam Damai
1   Ampang, Kuala Lumpur
2       Bandar Menjalara
3  Bandar Sri Permaisuri
4   Bandar Tasik Selatan

In [7]:
# print the number of rows of the dataframe
kl_df.shape

(71, 1)

### 3. Get the geographical coordinates

In [8]:
# define a function to get coordinates
def get_latlng(neighborhood):
    # initialize your variable to None
    lat_lng_coords = None
    # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, Kuala Lumpur, Malaysia'.format(neighborhood))
        lat_lng_coords = g.latlng
    return lat_lng_coords

In [9]:
# call the function to get the coordinates, store in a new list using list comprehension
coords = [ get_latlng(neighborhood) for neighborhood in kl_df["Neighborhood"].tolist() ]

In [10]:
coords

[[3.0576900000000364, 101.74388000000005],
 [3.1484988508598852, 101.69672774991264],
 [3.1903500000000236, 101.62545000000006],
 [3.1039100000000417, 101.71226000000007],
 [3.072750000000042, 101.71461000000005],
 [3.08280000000002, 101.72281000000004],
 [3.1292000000000257, 101.67844000000008],
 [3.1292000000000257, 101.67844000000008],
 [3.111020000000053, 101.66283000000004],
 [3.061870000000056, 101.74675000000008],
 [3.147890000000075, 101.69405000000006],
 [3.12916000000007, 101.68406000000004],
 [3.147770000000037, 101.70855000000006],
 [3.0578100000000745, 101.68965000000009],
 [3.1434800000000678, 101.64433000000008],
 [3.1486092228953293, 101.69985445868105],
 [3.129290000000026, 101.69896000000006],
 [3.17381000000006, 101.68276000000009],
 [3.061870000000056, 101.74675000000008],
 [3.163590000000056, 101.69811000000004],
 [3.1479700000000435, 101.66795000000008],
 [3.1364442625375997, 101.69029430612223],
 [3.138766317393615, 101.68401504688993],
 [3.1578250453611005, 101.

In [11]:
# create temporary dataframe to populate the coordinates into Latitude and Longitude
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])

In [12]:
# merge the coordinates into the original dataframe
kl_df['Latitude'] = df_coords['Latitude']
kl_df['Longitude'] = df_coords['Longitude']

In [13]:
# check the neighborhoods and the coordinates
print(kl_df.shape)
kl_df

(71, 3)


Neighborhood  Latitude   Longitude
0                     Alam Damai  3.057690  101.743880
1           Ampang, Kuala Lumpur  3.148499  101.696728
2               Bandar Menjalara  3.190350  101.625450
3          Bandar Sri Permaisuri  3.103910  101.712260
4           Bandar Tasik Selatan  3.072750  101.714610
5               Bandar Tun Razak  3.082800  101.722810
6                        Bangsar  3.129200  101.678440
7                   Bangsar Park  3.129200  101.678440
8                  Bangsar South  3.111020  101.662830
9                 Batu 11 Cheras  3.061870  101.746750
10            Batu, Kuala Lumpur  3.147890  101.694050
11                   Brickfields  3.129160  101.684060
12                 Bukit Bintang  3.147770  101.708550
13                   Bukit Jalil  3.057810  101.689650
14                   Bukit Kiara  3.143480  101.644330
15                   Bukit Nanas  3.148609  101.699854
16                Bukit Petaling  3.129290  101.698960
17                   Bukit Tunku  3.173810  101.682760
18          Cheras, Kuala Lumpur  3.061870  101.746750
19                      Chow Kit  3.163590  101.698110
20             Damansara Heights  3.147970  101.667950
21         Damansara Town Centre  3.136444  101.690294
22       Damansara, Kuala Lumpur  3.138766  101.684015
23                    Dang Wangi  3.157825  101.697280
24                 Desa Petaling  3.083300  101.704380
25    Federal Hill, Kuala Lumpur  3.136370  101.685640
26                  Happy Garden  3.201630  101.721070
27  Jalan Cochrane, Kuala Lumpur  3.134630  101.721690
28                    Jalan Duta  3.180025  101.677833
29                       Jinjang  3.209500  101.658740
30    Kampung Baru, Kuala Lumpur  3.165460  101.710280
31         Kampung Datuk Keramat  3.166400  101.730460
32         Kampung Padang Balang  3.209430  101.693180
33                        Kepong  3.217500  101.637630
34                   Kepong Baru  3.207771  101.645173
35                   KL Eco City  3.117130  101.673840
36                   Kuchai Lama  3.090740  101.677330
37                 Lembah Pantai  3.121189  101.663889
38                        Maluri  3.147890  101.694050
39                  Medan Tuanku  3.159260  101.698340
40                       Miharja  3.147890  101.694050
41                    Mont Kiara  3.165320  101.652430
42                  Pantai Dalam  3.094760  101.667470
43            Pudu, Kuala Lumpur  3.133540  101.713070
44                     Putrajaya  3.125851  101.718513
45                   Salak South  3.081540  101.696890
46                      Segambut  3.186390  101.668100
47                       Semarak  3.179943  101.721449
48          Sentul, Kuala Lumpur  3.175080  101.693050
49                       Setapak  3.188160  101.704150
50                   Setiawangsa  3.191802  101.740066
51                      Shamelin  3.124570  101.735970
52                  Sri Hartamas  3.162200  101.650360
53                  Sri Petaling  3.072600  101.682520
54                   Sungai Besi  3.049970  101.706030
55            Taman Bukit Maluri  3.200660  101.633370
56               Taman Connaught  3.082690  101.736890
57                    Taman Desa  3.102970  101.684710
58                 Taman Ibukota  3.212160  101.715400
59                Taman Len Seng  3.069080  101.742870
60                  Taman Melati  3.223570  101.723990
61                   Taman Midah  3.093590  101.728370
62                     Taman OUG  3.210051  101.634508
63               Taman P. Ramlee  3.193940  101.705730
64               Taman Sri Sinar  3.190070  101.652930
65            Taman Taynton View  3.087070  101.736810
66           Taman Tun Dr Ismail  3.152830  101.622710
67                 Taman U-Thant  3.157700  101.724520
68                   Taman Wahyu  3.222400  101.671730
69                    Titiwangsa  3.180730  101.703210
70                   Wangsa Maju  3.203870  101.737150

In [14]:
# save the DataFrame as CSV file
kl_df.to_csv("kl_df.csv", index=False)

### 4. Create a map of Kuala Lumpur with neighborhoods superimposed on top

In [15]:
# get the coordinates of Kuala Lumpur
address = 'Kuala Lumpur, Malaysia'

geolocator = Nominatim(user_agent="ahmadsyahmiadnan@gmail.com")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Kuala Lumpur, Malaysiae {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Kuala Lumpur, Malaysiae 3.1516964, 101.6942371.


In [16]:
# create map of Toronto using latitude and longitude values
map_kl = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, neighborhood in zip(kl_df['Latitude'], kl_df['Longitude'], kl_df['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_kl)  
    
map_kl

In [17]:
# save the map as HTML file
map_kl.save('map_kl.html')

### 5. Use the Foursquare API to explore the neighborhoods

In [18]:
# define Foursquare Credentials and Version
CLIENT_ID = 'FRXGPXSPI5K5IBO1NYCOD520LFQ1E1F0LODZLDCW3RBIGCPR' # your Foursquare ID
CLIENT_SECRET = '0LD4BP4G342T5C3UKZ2HLLEJTT3KN2INJLJVZJ11OZPLYPPW' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: FRXGPXSPI5K5IBO1NYCOD520LFQ1E1F0LODZLDCW3RBIGCPR
CLIENT_SECRET:0LD4BP4G342T5C3UKZ2HLLEJTT3KN2INJLJVZJ11OZPLYPPW


**Now, let's get the top 100 venues that are within a radius of 2000 meters.**

In [20]:
radius = 2000
LIMIT = 100

venues = []

for lat, long, neighborhood in zip(kl_df['Latitude'], kl_df['Longitude'], kl_df['Neighborhood']):
    
    # create the API request URL
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    # return only relevant information for each nearby venue
    for venue in results:
        venues.append((
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [21]:
# convert the venues list into a new DataFrame
venues_df = pd.DataFrame(venues)

# define the column names
venues_df.columns = ['Neighborhood', 'Latitude', 'Longitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']

print(venues_df.shape)
venues_df.head()

(7084, 7)


Neighborhood  Latitude  Longitude  \
0   Alam Damai   3.05769  101.74388   
1   Alam Damai   3.05769  101.74388   
2   Alam Damai   3.05769  101.74388   
3   Alam Damai   3.05769  101.74388   
4   Alam Damai   3.05769  101.74388   

                                       VenueName  VenueLatitude  \
0                  Pengedar Shaklee Kuala Lumpur       3.061235   
1                                   Jc Deli 皆喜食坊       3.058397   
2                               Machi Noodle 妈子面       3.057695   
3                             Roti Canai Perindu       3.055071   
4  Ivy Sekinchan Seafood Noodle House 适耕莊特制魚丸海鲜面       3.065749   

   VenueLongitude      VenueCategory  
0      101.740696    Supplement Shop  
1      101.748560  Food & Drink Shop  
2      101.746635       Noodle House  
3      101.740574     Breakfast Spot  
4      101.748718       Noodle House

In [22]:
venues_df[venues_df['VenueCategory']== 'Café'].shape #delete this later

(316, 7)

**Let's check how many venues were returned for each neighorhood**

In [23]:
venues_df.groupby(["Neighborhood"]).count()

Latitude  Longitude  VenueName  VenueLatitude  \
Neighborhood                                                                  
Alam Damai                         100        100        100            100   
Ampang, Kuala Lumpur               100        100        100            100   
Bandar Menjalara                   100        100        100            100   
Bandar Sri Permaisuri              100        100        100            100   
Bandar Tasik Selatan                97         97         97             97   
Bandar Tun Razak                   100        100        100            100   
Bangsar                            100        100        100            100   
Bangsar Park                       100        100        100            100   
Bangsar South                      100        100        100            100   
Batu 11 Cheras                     100        100        100            100   
Batu, Kuala Lumpur                 100        100        100            100   
Brickfields                        100        100        100            100   
Bukit Bintang                      100        100        100            100   
Bukit Jalil                        100        100        100            100   
Bukit Kiara                        100        100        100            100   
Bukit Nanas                        100        100        100            100   
Bukit Petaling                     100        100        100            100   
Bukit Tunku                        100        100        100            100   
Cheras, Kuala Lumpur               100        100        100            100   
Chow Kit                           100        100        100            100   
Damansara Heights                  100        100        100            100   
Damansara Town Centre              100        100        100            100   
Damansara, Kuala Lumpur            100        100        100            100   
Dang Wangi                         100        100        100            100   
Desa Petaling                      100        100        100            100   
Federal Hill, Kuala Lumpur         100        100        100            100   
Happy Garden                       100        100        100            100   
Jalan Cochrane, Kuala Lumpur       100        100        100            100   
Jalan Duta                         100        100        100            100   
Jinjang                            100        100        100            100   
KL Eco City                        100        100        100            100   
Kampung Baru, Kuala Lumpur         100        100        100            100   
Kampung Datuk Keramat              100        100        100            100   
Kampung Padang Balang               95         95         95             95   
Kepong                             100        100        100            100   
Kepong Baru                        100        100        100            100   
Kuchai Lama                        100        100        100            100   
Lembah Pantai                      100        100        100            100   
Maluri                             100        100        100            100   
Medan Tuanku                       100        100        100            100   
Miharja                            100        100        100            100   
Mont Kiara                         100        100        100            100   
Pantai Dalam                       100        100        100            100   
Pudu, Kuala Lumpur                 100        100        100            100   
Putrajaya                          100        100        100            100   
Salak South                        100        100        100            100   
Segambut                           100        100        100            100   
Semarak                            100        100        100            100   
Sentul, Kuala Lumpur               100        100        100            100   
S

**Let's find out how many unique categories can be curated from all the returned venues**

In [24]:
print('There are {} uniques categories.'.format(len(venues_df['VenueCategory'].unique())))

There are 304 uniques categories.


In [25]:
# print out the list of categories
venues_df['VenueCategory'].unique()[:50]

array(['Supplement Shop', 'Food & Drink Shop', 'Noodle House',
       'Breakfast Spot', 'Chinese Restaurant', 'Seafood Restaurant',
       'Restaurant', 'Food Court', 'Vegetarian / Vegan Restaurant',
       'Dessert Shop', 'Bubble Tea Shop', 'Park', 'Indian Restaurant',
       'Asian Restaurant', 'Coffee Shop', 'Japanese Restaurant',
       'Other Great Outdoors', 'Spa', 'Convenience Store', 'Pet Store',
       'Dim Sum Restaurant', 'Snack Place', 'Food Truck', 'Outlet Store',
       'Chinese Breakfast Place', 'Bakery', 'Café', 'Farmers Market',
       'Gym / Fitness Center', 'Malay Restaurant', 'Cantonese Restaurant',
       'Steakhouse', 'Fast Food Restaurant', 'Hakka Restaurant',
       'Badminton Court', 'Athletics & Sports', 'Mamak Restaurant',
       'Middle Eastern Restaurant', 'Burger Joint', 'Winery',
       'College Bookstore', 'Grocery Store', 'Diner', 'Electronics Store',
       'Gourmet Shop', 'Vietnamese Restaurant', 'Soup Place', 'Hostel',
       'South Indian Restaurant

In [26]:
# check if the results contain "Shopping Mall"
"Neighborhood" in venues_df['VenueCategory'].unique()

True

### 6. Analyze Each Neighborhood

In [27]:
# one hot encoding
kl_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
kl_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [kl_onehot.columns[-1]] + list(kl_onehot.columns[:-1])
kl_onehot = kl_onehot[fixed_columns]

print(kl_onehot.shape)
kl_onehot.head()

(7084, 305)


Neighborhoods  Accessories Store  Adult Boutique  African Restaurant  \
0    Alam Damai                  0               0                   0   
1    Alam Damai                  0               0                   0   
2    Alam Damai                  0               0                   0   
3    Alam Damai                  0               0                   0   
4    Alam Damai                  0               0                   0   

   American Restaurant  Arcade  Art Gallery  Art Museum  Arts & Crafts Store  \
0                    0       0            0           0                    0   
1                    0       0            0           0                    0   
2                    0       0            0           0                    0   
3                    0       0            0           0                    0   
4                    0       0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auditorium  Australian Restaurant  \
0                 0                   0           0                      0   
1                 0                   0           0                      0   
2                 0                   0           0                      0   
3                 0                   0           0                      0   
4                 0                   0           0                      0   

   Auto Garage  Auto Workshop  Automotive Shop  BBQ Joint  Badminton Court  \
0            0              0                0          0                0   
1            0              0                0          0                0   
2            0              0                0          0                0   
3            0              0                0          0                0   
4            0              0                0          0                0   

   Bagel Shop  Bakery  Bar  Baseball Field  Basketball Court  Beer Bar  \
0           0       0    0               0                 0         0   
1           0       0    0               0                 0         0   
2           0       0    0               0                 0         0   
3           0       0    0               0                 0         0   
4           0       0    0               0                 0         0   

   Beer Garden  Betting Shop  Bistro  Board Shop  Bookstore  Botanical Garden  \
0            0             0       0           0          0                 0   
1            0             0       0           0          0                 0   
2            0             0       0           0          0                 0   
3            0             0       0           0          0                 0   
4            0             0       0           0          0                 0   

   Boutique  Bowling Alley  Boxing Gym  Brazilian Restaurant  Breakfast Spot  \
0         0              0           0                     0               0   
1         0              0           0                     0               0   
2         0              0           0                     0               0   
3         0              0           0                     0               1   
4         0              0           0                     0               0   

   Brewery  Bridal Shop  Bubble Tea Shop  Buffet  Building  Burger Joint  \
0        0            0                0       0         0             0   
1        0            0                0       0         0             0   
2        0            0                0       0         0             0   
3        0            0                0       0         0             0   
4        0            0                0       0         0             0   

   Burrito Place  Business Service  Butcher  Café  Camera Store  Campground  \
0              0                 0        0     0             0           0   
1              0                 0        0     0             0           0   
2              0               

**Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category**

In [28]:
kl_grouped = kl_onehot.groupby(["Neighborhoods"]).mean().reset_index()

print(kl_grouped.shape)
kl_grouped

(71, 305)


Neighborhoods  Accessories Store  Adult Boutique  \
0                     Alam Damai           0.000000            0.00   
1           Ampang, Kuala Lumpur           0.000000            0.00   
2               Bandar Menjalara           0.000000            0.00   
3          Bandar Sri Permaisuri           0.010000            0.00   
4           Bandar Tasik Selatan           0.010309            0.00   
5               Bandar Tun Razak           0.010000            0.00   
6                        Bangsar           0.000000            0.00   
7                   Bangsar Park           0.000000            0.00   
8                  Bangsar South           0.010000            0.00   
9                 Batu 11 Cheras           0.000000            0.00   
10            Batu, Kuala Lumpur           0.000000            0.01   
11                   Brickfields           0.000000            0.00   
12                 Bukit Bintang           0.010000            0.00   
13                   Bukit Jalil           0.000000            0.00   
14                   Bukit Kiara           0.000000            0.00   
15                   Bukit Nanas           0.000000            0.00   
16                Bukit Petaling           0.000000            0.00   
17                   Bukit Tunku           0.000000            0.00   
18          Cheras, Kuala Lumpur           0.000000            0.00   
19                      Chow Kit           0.000000            0.01   
20             Damansara Heights           0.000000            0.00   
21         Damansara Town Centre           0.000000            0.00   
22       Damansara, Kuala Lumpur           0.000000            0.00   
23                    Dang Wangi           0.000000            0.00   
24                 Desa Petaling           0.000000            0.00   
25    Federal Hill, Kuala Lumpur           0.000000            0.00   
26                  Happy Garden           0.000000            0.00   
27  Jalan Cochrane, Kuala Lumpur           0.000000            0.00   
28                    Jalan Duta           0.000000            0.00   
29                       Jinjang           0.010000            0.00   
30                   KL Eco City           0.010000            0.00   
31    Kampung Baru, Kuala Lumpur           0.010000            0.00   
32         Kampung Datuk Keramat           0.000000            0.00   
33         Kampung Padang Balang           0.000000            0.00   
34                        Kepong           0.010000            0.00   
35                   Kepong Baru           0.010000            0.00   
36                   Kuchai Lama           0.000000            0.00   
37                 Lembah Pantai           0.010000            0.00   
38                        Maluri           0.000000            0.01   
39                  Medan Tuanku           0.000000            0.00   
40                       Miharja           0.000000            0.01   
41                    Mont Kiara           0.000000            0.00   
42                  Pantai Dalam           0.000000            0.00   
43            Pudu, Kuala Lumpur           0.000000            0.00   
44                     Putrajaya           0.000000            0.00   
45                   Salak South           0.000000            0.00   
46                      Segambut           0.000000            0.00   
47                       Semarak           0.000000            0.00   
48          Sentul, Kuala Lumpur           0.000000            0.00   
49                       Setapak           0.000000            0.00   
50                   Setiawangsa           0.000000            0.00   
51                      Shamelin           0.000000            0.00   
52                  Sri Hartamas           0.000000            0.00   
53                  Sri Petaling           0.000000            0.00   
54                   Sungai Besi           0.000000            0.00   
55            Taman Bukit Maluri           

In [29]:
kl_grouped['Coffee'] = kl_grouped['Coffee Shop'] + kl_grouped['Café'] #Combine 'Café' and 'Coffee Shop' to one column

In [30]:
len(kl_grouped[kl_grouped["Coffee"] > 0])

71

**Create a new DataFrame for Shopping Mall data only**

In [31]:
kl_coffee = kl_grouped[["Neighborhoods","Coffee"]]

In [32]:
kl_coffee.head()

Neighborhoods    Coffee
0             Alam Damai  0.050000
1   Ampang, Kuala Lumpur  0.080000
2       Bandar Menjalara  0.170000
3  Bandar Sri Permaisuri  0.010000
4   Bandar Tasik Selatan  0.030928

### 7. Cluster Neighborhoods
Run k-means to cluster the neighborhoods in Kuala Lumpur into 3 clusters.

In [33]:
# set number of clusters
kclusters = 3

kl_clustering = kl_coffee.drop(["Neighborhoods"], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(kl_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 0, 1, 2, 2, 2, 0, 0, 0, 2])

In [34]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.
kl_merged = kl_coffee.copy()

# add clustering labels
kl_merged["Cluster Labels"] = kmeans.labels_

In [35]:
kl_merged.rename(columns={"Neighborhoods": "Neighborhood"}, inplace=True)
kl_merged.head()

Neighborhood    Coffee  Cluster Labels
0             Alam Damai  0.050000               2
1   Ampang, Kuala Lumpur  0.080000               0
2       Bandar Menjalara  0.170000               1
3  Bandar Sri Permaisuri  0.010000               2
4   Bandar Tasik Selatan  0.030928               2

In [36]:
# Add latitude/longitude for each neighborhood
kl_merged = kl_merged.join(kl_df.set_index("Neighborhood"), on="Neighborhood")

print(kl_merged.shape)
kl_merged.head() # check the last columns!

(71, 5)


Neighborhood    Coffee  Cluster Labels  Latitude   Longitude
0             Alam Damai  0.050000               2  3.057690  101.743880
1   Ampang, Kuala Lumpur  0.080000               0  3.148499  101.696728
2       Bandar Menjalara  0.170000               1  3.190350  101.625450
3  Bandar Sri Permaisuri  0.010000               2  3.103910  101.712260
4   Bandar Tasik Selatan  0.030928               2  3.072750  101.714610

In [37]:
# sort the results by Cluster Labels
print(kl_merged.shape)
kl_merged.sort_values(["Cluster Labels"], inplace=True)
kl_merged

(71, 5)


Neighborhood    Coffee  Cluster Labels  Latitude  \
70                   Wangsa Maju  0.090000               0  3.203870   
62                     Taman OUG  0.070000               0  3.210051   
61                   Taman Midah  0.080000               0  3.093590   
51                      Shamelin  0.080000               0  3.124570   
59                Taman Len Seng  0.090000               0  3.069080   
30                   KL Eco City  0.070000               0  3.117130   
32         Kampung Datuk Keramat  0.100000               0  3.166400   
56               Taman Connaught  0.090000               0  3.082690   
34                        Kepong  0.080000               0  3.217500   
55            Taman Bukit Maluri  0.080000               0  3.200660   
36                   Kuchai Lama  0.090000               0  3.090740   
37                 Lembah Pantai  0.080000               0  3.121189   
54                   Sungai Besi  0.090000               0  3.049970   
42                  Pantai Dalam  0.080000               0  3.094760   
43            Pudu, Kuala Lumpur  0.080000               0  3.133540   
46                      Segambut  0.100000               0  3.186390   
47                       Semarak  0.070000               0  3.179943   
48          Sentul, Kuala Lumpur  0.100000               0  3.175080   
22       Damansara, Kuala Lumpur  0.090000               0  3.138766   
21         Damansara Town Centre  0.100000               0  3.136444   
25    Federal Hill, Kuala Lumpur  0.090000               0  3.136370   
50                   Setiawangsa  0.070000               0  3.191802   
20             Damansara Heights  0.070000               0  3.147970   
6                        Bangsar  0.080000               0  3.129200   
11                   Brickfields  0.080000               0  3.129160   
12                 Bukit Bintang  0.090000               0  3.147770   
8                  Bangsar South  0.080000               0  3.111020   
66           Taman Tun Dr Ismail  0.090000               0  3.152830   
15                   Bukit Nanas  0.090000               0  3.148609   
16                Bukit Petaling  0.090000               0  3.129290   
1           Ampang, Kuala Lumpur  0.080000               0  3.148499   
14                   Bukit Kiara  0.080000               0  3.143480   
7                   Bangsar Park  0.080000               0  3.129200   
53                  Sri Petaling  0.120000               1  3.072600   
52                  Sri Hartamas  0.110000               1  3.162200   
41                    Mont Kiara  0.130000               1  3.165320   
40                       Miharja  0.120000               1  3.147890   
2               Bandar Menjalara  0.170000               1  3.190350   
65            Taman Taynton View  0.110000               1  3.087070   
10            Batu, Kuala Lumpur  0.120000               1  3.147890   
57                    Taman Desa  0.110000               1  3.102970   
13                   Bukit Jalil  0.190000               1  3.057810   
28                    Jalan Duta  0.120000               1  3.180025   
17                   Bukit Tunku  0.110000               1  3.173810   
38                        Maluri  0.120000               1  3.147890   
58                 Taman Ibukota  0.020000               2  3.212160   
67                 Taman U-Thant  0.050000               2  3.157700   
60                  Taman Melati  0.030000               2  3.223570   
68                   Taman Wahyu  0.030000               2  3.222400   
64               Taman Sri Sinar  0.061224               2  3.190070   
63               Taman P. Ramlee  0.042553               2  3.193940   
0                     Alam Damai  0.050000               2  3.057690   
45                   Salak South  0.060000               2  3.081540   
3          Bandar Sri Permaisuri  0.010000               2  3.103910   
4           Bandar Tasik Selatan  0.030928               2  

**Finally, let's visualize the resulting clusters**

In [38]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(kl_merged['Latitude'], kl_merged['Longitude'], kl_merged['Neighborhood'], kl_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' - Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [39]:
# save the map as HTML file
map_clusters.save('map_clusters.html')

### 8. Examine Clusters

#### Cluster 0

In [49]:
cluster0 = kl_merged.loc[kl_merged['Cluster Labels'] == 0]
print(cluster0.shape)
cluster0

(33, 5)


Neighborhood  Coffee  Cluster Labels  Latitude   Longitude
70                 Wangsa Maju    0.09               0  3.203870  101.737150
62                   Taman OUG    0.07               0  3.210051  101.634508
61                 Taman Midah    0.08               0  3.093590  101.728370
51                    Shamelin    0.08               0  3.124570  101.735970
59              Taman Len Seng    0.09               0  3.069080  101.742870
30                 KL Eco City    0.07               0  3.117130  101.673840
32       Kampung Datuk Keramat    0.10               0  3.166400  101.730460
56             Taman Connaught    0.09               0  3.082690  101.736890
34                      Kepong    0.08               0  3.217500  101.637630
55          Taman Bukit Maluri    0.08               0  3.200660  101.633370
36                 Kuchai Lama    0.09               0  3.090740  101.677330
37               Lembah Pantai    0.08               0  3.121189  101.663889
54                 Sungai Besi    0.09               0  3.049970  101.706030
42                Pantai Dalam    0.08               0  3.094760  101.667470
43          Pudu, Kuala Lumpur    0.08               0  3.133540  101.713070
46                    Segambut    0.10               0  3.186390  101.668100
47                     Semarak    0.07               0  3.179943  101.721449
48        Sentul, Kuala Lumpur    0.10               0  3.175080  101.693050
22     Damansara, Kuala Lumpur    0.09               0  3.138766  101.684015
21       Damansara Town Centre    0.10               0  3.136444  101.690294
25  Federal Hill, Kuala Lumpur    0.09               0  3.136370  101.685640
50                 Setiawangsa    0.07               0  3.191802  101.740066
20           Damansara Heights    0.07               0  3.147970  101.667950
6                      Bangsar    0.08               0  3.129200  101.678440
11                 Brickfields    0.08               0  3.129160  101.684060
12               Bukit Bintang    0.09               0  3.147770  101.708550
8                Bangsar South    0.08               0  3.111020  101.662830
66         Taman Tun Dr Ismail    0.09               0  3.152830  101.622710
15                 Bukit Nanas    0.09               0  3.148609  101.699854
16              Bukit Petaling    0.09               0  3.129290  101.698960
1         Ampang, Kuala Lumpur    0.08               0  3.148499  101.696728
14                 Bukit Kiara    0.08               0  3.143480  101.644330
7                 Bangsar Park    0.08               0  3.129200  101.678440

#### Cluster 1

In [50]:
cluster1 = kl_merged.loc[kl_merged['Cluster Labels'] == 1]
print(cluster1.shape)
cluster1

(12, 5)


Neighborhood  Coffee  Cluster Labels  Latitude   Longitude
53        Sri Petaling    0.12               1  3.072600  101.682520
52        Sri Hartamas    0.11               1  3.162200  101.650360
41          Mont Kiara    0.13               1  3.165320  101.652430
40             Miharja    0.12               1  3.147890  101.694050
2     Bandar Menjalara    0.17               1  3.190350  101.625450
65  Taman Taynton View    0.11               1  3.087070  101.736810
10  Batu, Kuala Lumpur    0.12               1  3.147890  101.694050
57          Taman Desa    0.11               1  3.102970  101.684710
13         Bukit Jalil    0.19               1  3.057810  101.689650
28          Jalan Duta    0.12               1  3.180025  101.677833
17         Bukit Tunku    0.11               1  3.173810  101.682760
38              Maluri    0.12               1  3.147890  101.694050

#### Cluster 2

In [51]:
cluster2 = kl_merged.loc[kl_merged['Cluster Labels'] == 2]
print(cluster2.shape)
cluster2

(26, 5)


Neighborhood    Coffee  Cluster Labels  Latitude  \
58                 Taman Ibukota  0.020000               2  3.212160   
67                 Taman U-Thant  0.050000               2  3.157700   
60                  Taman Melati  0.030000               2  3.223570   
68                   Taman Wahyu  0.030000               2  3.222400   
64               Taman Sri Sinar  0.061224               2  3.190070   
63               Taman P. Ramlee  0.042553               2  3.193940   
0                     Alam Damai  0.050000               2  3.057690   
45                   Salak South  0.060000               2  3.081540   
3          Bandar Sri Permaisuri  0.010000               2  3.103910   
4           Bandar Tasik Selatan  0.030928               2  3.072750   
5               Bandar Tun Razak  0.060000               2  3.082800   
9                 Batu 11 Cheras  0.060000               2  3.061870   
18          Cheras, Kuala Lumpur  0.060000               2  3.061870   
19                      Chow Kit  0.050000               2  3.163590   
23                    Dang Wangi  0.030000               2  3.157825   
49                       Setapak  0.060000               2  3.188160   
24                 Desa Petaling  0.060000               2  3.083300   
27  Jalan Cochrane, Kuala Lumpur  0.050000               2  3.134630   
29                       Jinjang  0.040000               2  3.209500   
31    Kampung Baru, Kuala Lumpur  0.050000               2  3.165460   
33         Kampung Padang Balang  0.010526               2  3.209430   
69                    Titiwangsa  0.060000               2  3.180730   
39                  Medan Tuanku  0.030000               2  3.159260   
44                     Putrajaya  0.040000               2  3.125851   
26                  Happy Garden  0.060000               2  3.201630   
35                   Kepong Baru  0.060000               2  3.207771   

     Longitude  
58  101.715400  
67  101.724520  
60  101.723990  
68  101.671730  
64  101.652930  
63  101.705730  
0   101.743880  
45  101.696890  
3   101.712260  
4   101.714610  
5   101.722810  
9   101.746750  
18  101.746750  
19  101.698110  
23  101.697280  
49  101.704150  
24  101.704380  
27  101.721690  
29  101.658740  
31  101.710280  
33  101.693180  
69  101.703210  
39  101.698340  
44  101.718513  
26  101.721070  
35  101.645173

#### Observations: 
From the first observation, we can see that all neighbourhood have at least 1 coffee shop. 

**Cluster 0** shows the second highest density of cafés in the neighbourhoods which consist of *6 to 9 cafés*. This is the biggest cluster with total of 33 neighbourhoods and consist of almost half of all neighbourhoods in Kuala Lumpur (47%). Most of the neighbourhoods in this cluster are located at the central area of Kuala Lumpur which also consists a lot of commercial buildings and offices. Maybe the target market of cafés in that area is the worker from those commercial buildings.

**Cluster** 1 is the cluster of neighbourhood that have *more than 10 cafés*. This cluster consist of 12 neigbourhoods out of 70 neighbourhoods which equivalet of only 17% of neighbourhoods in Kuala Lumpur. From the Kuala Lumpur map, we can see that its located in area of Mont Kiara, Bukit Jalil, Bukit Tunku and Damansara which are categories as high end neighbourhood in Kuala Lumpur.

Last but not least, **Cluster 2** is the cluster that im looking for because of the low density of coffee shops which are *from 1 to 6 cafés* in each neighbourhood. This cluster consists of 22 neighbourhoods out of 70 neighbourhoods in Kuala Lumpur and most of the the neighbourhoods in this cluster located at northern and southeastern part of Kuala Lumpur. This area has high potentialfor opening of a new cae since there is less competitor compared to neighbourhoods in cluster 0 and 1.